---
title: "diffex_MIBC_v2"
output: html_document
date: "2023-05-16"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
setwd("/Users/jonathankahn/Documents/School/Research/Graim Lab/Bladder Cancer Mice Project/Analysis")
library(DESeq2)
library(tidyverse)
library(data.table)
library(tximport)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(ensembldb)
library(EnsDb.Hsapiens.v86)
library(org.Hs.eg.db)

# Import Metadata CSV
metadata <- fread("metadata.csv")
metadata$Condition <- factor(metadata$Condition)
metadata <- metadata[,-4]

# Import COSMIC Genes for Later
cosmic.genes <- read.table("COSMIC_bladder_cancer.ids",sep=',', header=FALSE, stringsAsFactors=FALSE)[,1]

# Create Sample Name Vectors (for later use)
allSamples <- c("C1","C2","C3","C4","C5","C6","D1","D2","D3","D4","D5","D6","D7","D8")
urineSamples <- c("C1","C2","C3","C4","C5","C6")
plasmaSamples <- c("D1","D2","D3","D4","D5","D6","D7","D8")

# Get Paths for TXImport
files <- file.path("quants_v3", list.files("quants_v3"), "quant.sf")
names(files) <- substring(list.files("quants_v3"),1,2) # removes the _quants

# For ENSEMBL library:
edb <- EnsDb.Hsapiens.v86
tx <- transcripts(edb)
tx2gene <- data.frame(tx$tx_id, tx$gene_id)
colnames(tx2gene) <- c("TXNAME", "GENEID")

# Load the Data
txi <- tximport(files, type = "salmon", tx2gene = tx2gene)
names(txi)
txi.tx <- tximport(files, type = "salmon", txOut = TRUE) # transcripts
txi.sum <- summarizeToGene(txi.tx, tx2gene) # for creating counts
all.equal(txi$counts, txi.sum$counts) # check if equivalent downstream
```

```{r}
# Counts Matrix and General Processing
dat <- as.data.frame(txi.sum$counts)

# HUGO Name Conversion
library(org.Hs.eg.db)
annot.df <- data.frame("Symbols" = mapIds(org.Hs.eg.db, keys = rownames(dat), column = "SYMBOL", keytype = "ENSEMBL"), dat)
annot.df$Symbols <- toupper(annot.df$Symbols)
annot.df <- annot.df[!duplicated(annot.df$Symbols),] # TODO: Think of Alternative Method with this Data Format (median somehow?)
annot.df <- annot.df[complete.cases(annot.df),]
rownames(annot.df) <- annot.df$Symbols
annot.df <- annot.df[,-1]
dat <- annot.df 

# Adding a Sum Counts Column
dat$Sum <- apply(dat, MARGIN = 1, FUN = sum, na.rm = TRUE)

# Removing ~ 5636 genes that had less than 10 counts total across all the samples
    # And Removing Sum Column
counts <- dat %>%
  filter(Sum >= 10) %>%
  select(-Sum) %>% round()

# Note genes removed for later purposes
removed_data <- dat %>%
  filter(Sum < 10) %>%
  select(-Sum) %>% round()

# Create Log2 Scaled Version for Later
countsLogScaled <- log2(counts+1)

```

```{r}
# T-SNE Plot
library(M3C)

tsne_result <- tsne(countsLogScaled, perplex = 4)
tsne_result$coordinates
tsne_df <- data.frame(tsne_result$data, Type = metadata$SpecimenType, Condition = metadata$Condition)

# Create a ggplot2 plot and customize the appearance of the points
ggplot(tsne_df, aes(x = X1, y = X2, color = Type, shape = Condition)) +
  geom_point(size = 5) +
  scale_shape_manual(values = c(16, 17, 15, 14)) +
  theme_classic()
```

```{r fig.height=4}
# ALL DATA

# Create DESeqDataSet from Tximport, Run DESeq, Results
dds <- DESeqDataSetFromTximport(txi,colData = metadata,design = ~ Condition)
keep <- rowSums(counts(dds)) >= 10 # Filter rows with less than 10 total counts
dds <- dds[keep,]
dds <- DESeq(dds)
res <- results(dds)

# Convert to HUGO
annot.df <- data.frame("Symbols" = mapIds(org.Hs.eg.db, keys = rownames(res), column = "SYMBOL", keytype = "ENSEMBL"), res)
annot.df$Symbols <- toupper(annot.df$Symbols)
annot.df <- annot.df[!duplicated(annot.df$Symbols),] # TODO: FIND BETTER WAY LATER
annot.df <- annot.df[complete.cases(annot.df),]
rownames(annot.df) <- annot.df$Symbols
annot.df <- annot.df[,-1]
res <- annot.df
summary(res)

# Volcano Plot
library(EnhancedVolcano)

EnhancedVolcano::EnhancedVolcano(res,lab = rownames(res),x = 'log2FoldChange',y = 'padj',title = 'All Samples',pCutoff = 0.1,FCcutoff = 1,pointSize = 3.0,labSize = 6.0)
#png('Volcano_AllSamples.png')

```


```{r fig.height=4}
# # # URINE DATA # # #
### Generate TXI for Just Urine Data ###

# Filter Metadata
samples_urine <- metadata[Sample %in% urineSamples,]
samples_urine$Condition <- factor(samples_urine$Condition)

# Grabbing File Paths
files <- file.path("quants_urine", list.files("quants_urine"), "quant.sf")
names(files) <- substring(list.files("quants_urine"),1,2) # removes the _quants

# Load the Data
txiUrine <- tximport(files, type = "salmon", tx2gene = tx2gene)
names(txi)
dds_urine <- DESeqDataSetFromTximport(txiUrine,colData = samples_urine,design = ~ Condition)
dds_urine <- DESeq(dds_urine)
res_urine <- results(dds_urine)

# Convert to HUGO
urine.df <- data.frame("Symbols" = mapIds(org.Hs.eg.db, keys = rownames(res_urine), column = "SYMBOL", keytype = "ENSEMBL"), res_urine)
urine.df$Symbols <- toupper(urine.df$Symbols)
urine.df <- urine.df[!duplicated(urine.df$Symbols),] # TODO: FIND BETTER WAY LATER
urine.df <- urine.df[complete.cases(urine.df),]
rownames(urine.df) <- urine.df$Symbols
urine.df <- urine.df[,-1]
res_urine <- urine.df
summary(res_urine)

# Volcano Plot
library(EnhancedVolcano)

EnhancedVolcano::EnhancedVolcano(res_urine,lab = rownames(res_urine),x = 'log2FoldChange',y = 'padj',title = 'Urine Samples',pCutoff = 0.1,FCcutoff = 1,pointSize = 3.0,labSize = 6.0)
#png('Volcano_AllSamples.png')

```
```{r fig.height=4}
# # # PLASMA DATA # # #
### Generate TXI for Just Plasma Data ###

# Filter Metadata
samples_plasma <- metadata[Sample %in% plasmaSamples,]
samples_plasma$Condition <- factor(samples_plasma$Condition)

# Grabbing File Paths
files <- file.path("quants_plasma", list.files("quants_plasma"), "quant.sf")
names(files) <- substring(list.files("quants_plasma"),1,2) # removes the _quants

# Load the Data
txiPlasma <- tximport(files, type = "salmon", tx2gene = tx2gene)
names(txi)
dds_plasma <- DESeqDataSetFromTximport(txiPlasma,colData = samples_plasma,design = ~ Condition)
dds_plasma <- DESeq(dds_plasma)
res_plasma <- results(dds_plasma)

# Convert to HUGO
plasma.df <- data.frame("Symbols" = mapIds(org.Hs.eg.db, keys = rownames(res_plasma), column = "SYMBOL", keytype = "ENSEMBL"), res_plasma)
plasma.df$Symbols <- toupper(plasma.df$Symbols)
plasma.df <- plasma.df[!duplicated(plasma.df$Symbols),] # TODO: FIND BETTER WAY LATER
plasma.df <- plasma.df[complete.cases(plasma.df),]
rownames(plasma.df) <- plasma.df$Symbols
plasma.df <- plasma.df[,-1]
res_plasma <- plasma.df
summary(res_plasma)

# Volcano Plot
library(EnhancedVolcano)

EnhancedVolcano::EnhancedVolcano(res_plasma,lab = rownames(res_plasma),x = 'log2FoldChange',y = 'padj',title = 'Plasma Samples',pCutoff = 0.1,FCcutoff = 1,pointSize = 3.0,labSize = 6.0)
#png('Volcano_AllSamples.png')

```

```{r}
# Additional Plot Generation
library(reshape2)
library(ggplot2)
corrs <- cor(counts, method = c("spearman")) 
corrs.melt <- melt(corrs)
ggplot(corrs.melt, aes(Var1, Var2, fill=value)) + geom_tile() + scale_fill_distiller(palette = "RdYlBu")
ggplot(corrs.melt, aes(Var1, Var2, fill=value)) + geom_tile() + scale_fill_gradient(low="white", high="steelblue", limits=c(0,1))  # all data

# COSMIC bladder cancer genes 
genes <- cosmic.genes
corrs <- cor(counts[genes,], method = c("spearman")) 
corrs.melt <- melt(corrs)
ggplot(corrs.melt, aes(Var1, Var2, fill=value)) + geom_tile() + scale_fill_distiller(palette = "RdYlBu",limits=c(-1,1))   # correlation heatmap
ggsave('HeatmapSpearmanCorrelation_COSMIC_BladderCancerGenesOnly.png',height=7.77,width=9.37)

dat.cosmic <- log2(as.matrix(counts[genes,])+1)
dat.melt <- melt(dat.cosmic)
colnames(dat.melt) <- c('Gene','Sample','value')
ggplot(dat.melt, aes(Gene, Sample, fill=value)) + geom_tile() + scale_fill_distiller(palette = "RdYlBu",limits=c(-6,6))   # expression heatmap
ggsave('Heatmap_COSMIC_BladderCancerGenesOnly.png',height=7.77,width=9.37)

```
```{r}
# Heatmap 
library(ComplexHeatmap)
column_ha <- HeatmapAnnotation(Specimen=metadata$SpecimenType,Condition=metadata$Condition, col=list(Specimen = c('Urine' = 'darkgreen','Plasma' = 'darkred'), Condition = c('NonInvasive' = "gray", 'Invasive' = "purple", 'AfterTreatment' = "green", 'Healthy' = "red")))
hm <- Heatmap(dat.cosmic, name = "Log2 Expr", bottom_annotation = column_ha)
draw(hm,
   column_title="COSMIC Bladder Cancer Genes",
   column_title_gp=grid::gpar(fontsize=16))
# Break into 2 heatmaps by specimen type
column_urine_ha <- HeatmapAnnotation(Specimen=metadata$SpecimenType[metadata$SpecimenType=="Urine"],Condition=metadata$Condition[metadata$SpecimenType=="Urine"], col=list(Specimen = c('Urine' = 'darkgreen'),Condition = c("NonInvasive" = "gray", "Invasive" = "purple")))

column_plasma_ha <- HeatmapAnnotation(Specimen=metadata$SpecimenType[metadata$SpecimenType=="Plasma"],Condition=metadata$Condition[metadata$SpecimenType=="Plasma"],col=list(Specimen = c('Plasma' = 'darkred'),Condition = c('NonInvasive' = "gray", 'Invasive' = "purple", 'AfterTreatment' = "green", 'Healthy' = "red")))

ht1 <- Heatmap(dat.cosmic[,metadata$SpecimenType=='Plasma'], name='Log2 Expr', bottom_annotation = column_plasma_ha)
ht2 <- Heatmap(dat.cosmic[,metadata$SpecimenType=='Urine'], name='Log2 Expr', bottom_annotation = column_urine_ha)
hm2 <- ht1 + ht2

draw(hm2, column_title = "COSMIC Genes Heatmap", merge_legends=TRUE) # add space for titles

# Heatmaps of the plasma and urine differentially expressed genes
#genes <- c("NIPAL2","ORC4","ARID3A","STX12","SGIP1","PTGFRN","PLPP2","LRP5","TRIM44","ACBD3","HACD4","GRIN3A","ADAM12","VANGL1")
#dat.exoDE <- log2(as.matrix(counts[genes,])+1)
#le <- metadata$condition
#ht1 <- Heatmap(dat.exoDE[,rownames(metadata)[metadata$SpecimenType=='Plasma']], name='Plasma',width=unit(8,"cm"))
#ht2 <- Heatmap(dat.exoDE[,rownames(samples)[metadata$SpecimenType=='Urine']], name='Urine',width=unit(6,"cm"), title='Urine')
#ht3 <- Heatmap(le, name = "Condition", width = unit(1, "cm"))
#ht1 + ht2 + ht3

#png(file="Heatmap_Annotated_ExoDiffExpr_Genes.png")
#draw(ht1 + ht2 + ht3, padding = unit(c(2, 2, 4, 2), "cm")) # add space for titles
#dev.off()

```






